filestore/util.go

```diff
@@ -0,0 +1,145 @@
+package filestore
+
+import (
+	"fmt"
+	//"io"
+	//"os"
+	//"path/filepath"
+
+	//"github.com/ipfs/go-ipfs/blocks"
+	//"github.com/ipfs/go-ipfs/blocks/blockstore"
+	pb "github.com/ipfs/go-ipfs/filestore/pb"
+	dshelp "github.com/ipfs/go-ipfs/thirdparty/ds-help"
+	//posinfo "github.com/ipfs/go-ipfs/thirdparty/posinfo"
+
+	ds "gx/ipfs/QmRWDav6mzWseLWeYfVd5fvUKiVe9xNH29YfMF438fG364/go-datastore"
+	//dsns "gx/ipfs/QmRWDav6mzWseLWeYfVd5fvUKiVe9xNH29YfMF438fG364/go-datastore/namespace"
+	dsq "gx/ipfs/QmRWDav6mzWseLWeYfVd5fvUKiVe9xNH29YfMF438fG364/go-datastore/query"
+	proto "gx/ipfs/QmT6n4mspWYEya864BhCUJEgyxiRfmiSY9ruQwTUNpRKaM/protobuf/proto"
+	cid "gx/ipfs/QmcTcsTvfaeEBRFo1TkFgT8sRmgi1n1LTZpecfVP8fzpGD/go-cid"
+)
+
+type Status int32
+
+const (
+	StatusOk        Status = 0
+	StatusFileError Status = 10 // Backing File Error
+	//StatusFileNotFound Status = 11 // Backing File Not Found
+	//StatusFileChanged  Status = 12 // Contents of the file changed
+	StatusOtherError Status = 20 // Internal Error, likely corrupt entry
+)
+
+func (s Status) String() string {
+	switch s {
+	case StatusOk:
+		return "ok"
+	case StatusFileError:
+		return "error"
+	case StatusOtherError:
+		return "ERROR"
+	default:
+		return "???"
+	}
+}
+
+func (s Status) Format() string {
+	return fmt.Sprintf("%-5s", s.String())
+}
+
+type ListRes struct {
+	Status   Status
+	ErrorMsg string
+	Key      *cid.Cid
+	FilePath string
+	Offset   uint64
+	Size     uint64
+}
+
+func (r *ListRes) FormatLong() string {
+	switch {
+	case r.Key == nil:
+		return "?????????????????????????????????????????????????"
+	default:
+		return fmt.Sprintf("%-50s %6d %s %d", r.Key, r.Size, r.FilePath, r.Offset)
+	}
+}
+
+func ListAll(fs *Filestore) (func () *ListRes, error) {
+	return listAll(fs, false)
+}
+
+func VerifyAll(fs *Filestore) (func () *ListRes, error) {
+	return listAll(fs, true)
+}
+
+func listAll(fs *Filestore, verify bool) (func () *ListRes, error) {
+	q := dsq.Query{}
+	qr, err := fs.fm.ds.Query(q)
+	if err != nil {
+		return nil, err
+	}
+
+	return func() *ListRes {
+		cid, dobj, err := next(qr)
+		if dobj == nil && err == nil {
+			return nil
+		} else if err == nil && verify {
+			_, err = fs.fm.readDataObj(cid, dobj)
+		}
+		return mkListRes(cid, dobj, err)
+	}, nil
+}
+
+func next(qr dsq.Results) (*cid.Cid, *pb.DataObj, error) {
+	v, ok := qr.NextSync()
+	if !ok {
+		return nil, nil, nil
+	}
+
+	k := ds.RawKey(v.Key)
+	c, err := dshelp.DsKeyToCid(k)
```

Comment:
Reusing that method will involve an extra lookup as we already have the value from the iterator.